core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -562,28 +561,38 @@ class KafkaApis(val requestChannel: RequestChannel,
             None
         }
 
-        downConvertMagic.map { magic =>
-          trace(s"Down converting records from partition $tp to message format version $magic for fetch request from $clientId")
-
-          // Because down-conversion is extremely memory intensive, we want to try and delay the down-conversion as much
-          // as possible. With KIP-283, we have the ability to lazily down-convert in a chunked manner. The lazy, chunked
-          // down-conversion always guarantees that at least one batch of messages is down-converted and sent out to the
-          // client.
-          new LazyDownConversionRecords(tp, unconvertedRecords, magic, fetchContext.getFetchOffset(tp).get, time)
-        }
-      }.getOrElse(unconvertedRecords)
+      // For fetch requests from clients, check if down-conversion is disabled for the particular partition
+      if (downConvertMagic.isDefined && !fetchRequest.isFromFollower &&
+          !logConfig.map(_.messageDownConversionEnable.booleanValue()).getOrElse(true)) {
```

Comment:
I think this can be simplified: `!logConfig.forall(_.messageDownConversionEnable)`?